File: server/src/api/cameras.rs
use std::sync::Arc;
use std::time::{SystemTime, UNIX_EPOCH};

use axum::extract::{Path, State};
use axum::http::StatusCode;
use axum::response::{IntoResponse, Response};
use axum::Extension;
use axum::Json;
use ghostcam::types::DeviceId;
use serde::{Deserialize, Serialize};

use super::auth::AuthUser;
use super::state::AppState;
use crate::db_trait::CameraUpdate;

#[derive(Serialize)]
pub struct CameraResponse {
    pub device_id: String,
    pub display_name: String,
    pub enrolled_at: u64,
    pub notes: Option<String>,
}

#[derive(Serialize)]
pub struct EnrollResponse {
    pub token: String,
    pub expires_at: u64,
}

/// GET /api/v1/cameras
pub async fn list(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<AuthUser>,
) -> Response {
    let cameras = match state.db.list_cameras(&user.user_id).await {
        Ok(c) => c,
        Err(e) => {
            tracing::error!("list cameras error: {e}");
            return StatusCode::INTERNAL_SERVER_ERROR.into_response();
        }
    };

    let responses: Vec<CameraResponse> = cameras
        .into_iter()
        .map(|cam| CameraResponse {
            device_id: cam.device_id.0,
            display_name: cam.display_name,
            enrolled_at: cam.enrolled_at,
            notes: cam.notes,
        })
        .collect();

    Json(responses).into_response()
}

/// POST /api/v1/cameras
pub async fn enroll(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<AuthUser>,
    Json(_body): Json<EnrollRequest>,
) -> Response {
    // Check camera limit (billing enforcement)
    match crate::billing::enforcement::check_camera_limit(
        state.db.as_ref(),
        &user.user_id,
        &state.tiers,
        state.stripe.is_some(),
    )
    .await
    {
        Ok(Ok(())) => {} // allowed
        Ok(Err(e)) => {
            state
                .audit
                .log(crate::audit::AuditEvent::CameraLimitBlocked {
                    user_id: user.user_id.0.clone(),
                    current: match &e {
                        crate::billing::enforcement::EnforcementError::CameraLimitReached {
                            current,
                            ..
                        } => *current,
                        _ => 0,
                    },
                    limit: match &e {
                        crate::billing::enforcement::EnforcementError::CameraLimitReached {
                            limit,
                            ..
                        } => *limit,
                        _ => 0,
                    },
                });
            let error_key = match &e {
                crate::billing::enforcement::EnforcementError::SubscriptionSuspended => {
                    "subscription_suspended"
                }
                _ => "camera_limit_reached",
            };
            return (
                StatusCode::PAYMENT_REQUIRED,
                Json(serde_json::json!({
                    "error": error_key,
                    "message": e.to_string(),
                })),
            )
                .into_response();
        }
        Err(e) => {
            tracing::error!("camera limit check failed: {e}");
            return StatusCode::INTERNAL_SERVER_ERROR.into_response();
        }
    }

    // Generate a one-time provision token
    let raw_token = crate::auth::generate_random_password();
    let token_hash = crate::auth::hmac_token(&raw_token, &state.hmac_secret);

    let now = SystemTime::now()
        .duration_since(UNIX_EPOCH)
        .unwrap()
        .as_secs();
    let expires_at = now + ghostcam::config::PROVISION_TOKEN_TTL_SECS;

    if let Err(e) = state
        .db
        .create_provision_token(&token_hash, &user.user_id, expires_at)
        .await
    {
        tracing::error!("failed to create provision token: {e}");
        return StatusCode::INTERNAL_SERVER_ERROR.into_response();
    }

    state
        .audit
        .log(crate::audit::AuditEvent::EnrollmentStarted {
            device_id: format!("provision-{}", &token_hash[..8]),
            owner_id: user.user_id.0.clone(),
        });

    Json(EnrollResponse {
        token: raw_token,
        expires_at,
    })
    .into_response()
}

/// Legacy request body for POST /api/v1/cameras.
/// Fields are accepted for backward compat (Docker entrypoint) but no longer
/// embedded in the JWT -- claim tokens are now stateless with just `sub`.
#[derive(Deserialize)]
#[allow(dead_code)]
pub struct EnrollRequest {
    pub display_name: Option<String>,
    pub wifi: Option<Vec<WifiCredential>>,
}

#[derive(Deserialize, Serialize)]
#[allow(dead_code)]
pub struct WifiCredential {
    pub ssid: String,
    pub psk: String,
}

/// GET /api/v1/cameras/:device_id
pub async fn get(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<AuthUser>,
    Path(device_id): Path<String>,
) -> Response {
    let device_id = DeviceId(device_id);
    let camera = match state.db.get_camera(&device_id).await {
        Ok(Some(c)) => c,
        Ok(None) => return StatusCode::NOT_FOUND.into_response(),
        Err(e) => {
            tracing::error!("get camera error: {e}");
            return StatusCode::INTERNAL_SERVER_ERROR.into_response();
        }
    };

    // Verify ownership
    if camera.user_id.as_ref() != Some(&user.user_id) {
        return StatusCode::NOT_FOUND.into_response();
    }

    Json(CameraResponse {
        device_id: camera.device_id.0,
        display_name: camera.display_name,
        enrolled_at: camera.enrolled_at,
        notes: camera.notes,
    })
    .into_response()
}

#[derive(Deserialize)]
pub struct UpdateCameraRequest {
    pub display_name: Option<String>,
    pub notes: Option<String>,
}

/// PATCH /api/v1/cameras/:device_id
pub async fn update(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<AuthUser>,
    Path(device_id): Path<String>,
    Json(body): Json<UpdateCameraRequest>,
) -> Response {
    let device_id = DeviceId(device_id);

    // Verify ownership
    match state.db.get_camera(&device_id).await {
        Ok(Some(c)) if c.user_id.as_ref() == Some(&user.user_id) => {}
        Ok(Some(_)) | Ok(None) => return StatusCode::NOT_FOUND.into_response(),
        Err(_) => return StatusCode::INTERNAL_SERVER_ERROR.into_response(),
    }

    let update = CameraUpdate {
        display_name: body.display_name,
        notes: body.notes,
    };

    if let Err(e) = state.db.update_camera(&device_id, &update).await {
        tracing::error!("update camera error: {e}");
        return StatusCode::INTERNAL_SERVER_ERROR.into_response();
    }

    StatusCode::OK.into_response()
}

/// DELETE /api/v1/cameras/:device_id
pub async fn delete(
    State(state): State<Arc<AppState>>,
    Extension(user): Extension<AuthUser>,
    Path(device_id): Path<String>,
) -> Response {
    let device_id = DeviceId(device_id);

    // Verify ownership
    match state.db.get_camera(&device_id).await {
        Ok(Some(c)) if c.user_id.as_ref() == Some(&user.user_id) => {}
        Ok(Some(_)) | Ok(None) => return StatusCode::NOT_FOUND.into_response(),
        Err(_) => return StatusCode::INTERNAL_SERVER_ERROR.into_response(),
    }

    if let Err(e) = state.db.delete_camera(&device_id).await {
        tracing::error!("delete camera error: {e}");
        return StatusCode::INTERNAL_SERVER_ERROR.into_response();
    }

    state
        .audit
        .log(crate::audit::AuditEvent::CameraUnregistered {
            device_id: device_id.0,
            initiated_by: user.user_id.0,
        });

    StatusCode::OK.into_response()
}

File: server/handlers/cameras.go
package handlers

import (
	"encoding/json"
	"log/slog"
	"net/http"
	"time"

	"github.com/cargocam/ghostcam/server/auth"
	"github.com/cargocam/ghostcam/server/billing"
	"github.com/cargocam/ghostcam/server/ctxutil"
	"github.com/cargocam/ghostcam/server/db"
	"github.com/go-chi/chi/v5"
)

const provisionTokenTTLSecs = 24 * 3600 // 24 hours

type cameraResponse struct {
	DeviceID      string  `json:"device_id"`
	DisplayName   string  `json:"display_name"`
	EnrolledAt    uint64  `json:"enrolled_at"`
	LastSeenAt    *int64  `json:"last_seen_at,omitempty"`
	Provisioned   bool    `json:"provisioned"`
	Notes         *string `json:"notes,omitempty"`
	Resolution    string  `json:"resolution"`
	RecordingMode string  `json:"recording_mode"`
}

type enrollResponse struct {
	Token     string `json:"token"`
	ExpiresAt uint64 `json:"expires_at"`
}

// ListCameras handles GET /api/v1/cameras.
func (h *Handlers) ListCameras(w http.ResponseWriter, r *http.Request) {
	userID := ctxutil.GetUserID(r)

	cameras, err := h.DB.ListCameras(r.Context(), userID)
	if err != nil {
		slog.Error("list cameras failed", "error", err)
		http.Error(w, "", http.StatusInternalServerError)
		return
	}

	resp := make([]cameraResponse, 0, len(cameras))
	for _, c := range cameras {
		resp = append(resp, cameraResponse{
			DeviceID:      c.DeviceID,
			DisplayName:   c.DisplayName,
			EnrolledAt:    uint64(c.EnrolledAt),
			LastSeenAt:    c.LastSeenAt,
			Provisioned:   c.LastSeenAt != nil,
			Notes:         c.Notes,
			Resolution:    c.Resolution,
			RecordingMode: c.RecordingMode,
		})
	}
	writeJSON(w, http.StatusOK, resp)
}

// Enroll handles POST /api/v1/cameras (generate provision token).
func (h *Handlers) Enroll(w http.ResponseWriter, r *http.Request) {
	userID := ctxutil.GetUserID(r)
	ctx := r.Context()

	// Enforce camera limit based on tier
	sub, _ := h.DB.GetSubscription(ctx, userID)
	tier := billing.GetTier(effectiveTier(sub, h.Stripe.SecretKey != ""))
	if tier.CameraLimit != nil {
		count, err := h.DB.GetCameraCount(ctx, userID)
		if err != nil {
			slog.Error("enroll: get camera count failed", "error", err)
			http.Error(w, "", http.StatusInternalServerError)
			return
		}
		if count >= int64(*tier.CameraLimit) {
			writeError(w, http.StatusPaymentRequired, "camera_limit_reached")
			return
		}
	}

	// Consume request body (ignored but accepted for backward compat)
	var body json.RawMessage
	_ = json.NewDecoder(r.Body).Decode(&body)

	rawToken := auth.GenerateRandomPassword()
	tokenHash := auth.HMACToken(rawToken, h.HMACSecret)

	now := uint64(time.Now().Unix())
	expiresAt := now + provisionTokenTTLSecs

	if err := h.DB.CreateProvisionToken(r.Context(), tokenHash, userID, int64(expiresAt)); err != nil {
		slog.Error("enroll: create provision token failed", "error", err)
		http.Error(w, "", http.StatusInternalServerError)
		return
	}

	db.AuditLog("enrollment_started", "user_id", userID)
	writeJSON(w, http.StatusOK, enrollResponse{Token: rawToken, ExpiresAt: expiresAt})
}

// GetCamera handles GET /api/v1/cameras/{deviceID}.
func (h *Handlers) GetCamera(w http.ResponseWriter, r *http.Request) {
	userID := ctxutil.GetUserID(r)
	deviceID := chi.URLParam(r, "deviceID")

	camera, err := h.DB.GetCamera(r.Context(), deviceID)
	if err != nil {
		slog.Error("get camera failed", "error", err)
		http.Error(w, "", http.StatusInternalServerError)
		return
	}
	if camera == nil || camera.UserID == nil || *camera.UserID != userID {
		http.Error(w, "", http.StatusNotFound)
		return
	}

	writeJSON(w, http.StatusOK, cameraResponse{
		DeviceID:      camera.DeviceID,
		DisplayName:   camera.DisplayName,
		EnrolledAt:    uint64(camera.EnrolledAt),
		LastSeenAt:    camera.LastSeenAt,
		Provisioned:   camera.LastSeenAt != nil,
		Notes:         camera.Notes,
		Resolution:    camera.Resolution,
		RecordingMode: camera.RecordingMode,
	})
}

type updateCameraRequest struct {
	DisplayName   *string `json:"display_name,omitempty"`
	Notes         *string `json:"notes,omitempty"`
	Resolution    *string `json:"resolution,omitempty"`
	RecordingMode *string `json:"recording_mode,omitempty"`
}

// UpdateCamera handles PATCH /api/v1/cameras/{deviceID}.
func (h *Handlers) UpdateCamera(w http.ResponseWriter, r *http.Request) {
	userID := ctxutil.GetUserID(r)
	deviceID := chi.URLParam(r, "deviceID")

	// Verify ownership
	camera, err := h.DB.GetCamera(r.Context(), deviceID)
	if err != nil || camera == nil || camera.UserID == nil || *camera.UserID != userID {
		http.Error(w, "", http.StatusNotFound)
		return
	}

	var body updateCameraRequest
	if err := json.NewDecoder(r.Body).Decode(&body); err != nil {
		writeError(w, http.StatusBadRequest, "invalid request body")
		return
	}

	// Validate resolution
	if body.Resolution != nil {
		switch *body.Resolution {
		case "480p", "720p", "1080p":
		default:
			writeError(w, http.StatusBadRequest, "resolution must be 480p, 720p, or 1080p")
			return
		}
	}

	// Validate recording mode
	if body.RecordingMode != nil {
		switch *body.RecordingMode {
		case "constant", "motion":
		default:
			writeError(w, http.StatusBadRequest, "recording_mode must be constant or motion")
			return
		}
	}

	ctx := r.Context()

	if err := h.DB.UpdateCamera(ctx, deviceID, &db.CameraUpdate{
		DisplayName:   body.DisplayName,
		Notes:         body.Notes,
		Resolution:    body.Resolution,
		RecordingMode: body.RecordingMode,
	}); err != nil {
		slog.Error("update camera failed", "error", err)
		http.Error(w, "", http.StatusInternalServerError)
		return
	}

	// Enqueue commands for settings that require camera restart
	if body.Resolution != nil && *body.Resolution != camera.Resolution {
		cmd, _ := json.Marshal(map[string]string{"type": "set_resolution", "resolution": *body.Resolution})
		if err := h.DB.EnqueueCommand(ctx, deviceID, cmd); err != nil {
			slog.Error("enqueue set_resolution failed", "error", err)
		}
	}
	if body.RecordingMode != nil && *body.RecordingMode != camera.RecordingMode {
		cmd, _ := json.Marshal(map[string]string{"type": "set_recording_mode", "mode": *body.RecordingMode})
		if err := h.DB.EnqueueCommand(ctx, deviceID, cmd); err != nil {
			slog.Error("enqueue set_recording_mode failed", "error", err)
		}
	}

	w.WriteHeader(http.StatusOK)
}

// DeleteCamera handles DELETE /api/v1/cameras/{deviceID}.
func (h *Handlers) DeleteCamera(w http.ResponseWriter, r *http.Request) {
	userID := ctxutil.GetUserID(r)
	deviceID := chi.URLParam(r, "deviceID")

	camera, err := h.DB.GetCamera(r.Context(), deviceID)
	if err != nil || camera == nil || camera.UserID == nil || *camera.UserID != userID {
		http.Error(w, "", http.StatusNotFound)
		return
	}

	if err := h.DB.DeleteCamera(r.Context(), deviceID); err != nil {
		slog.Error("delete camera failed", "error", err)
		http.Error(w, "", http.StatusInternalServerError)
		return
	}

	db.AuditLog("camera_unregistered", "device_id", deviceID, "initiated_by", userID)
	w.WriteHeader(http.StatusOK)
}
